fggs/multi.py

```diff
@@ -175,6 +175,55 @@ def multi_mv(a: MultiTensor, b: MultiTensor, transpose: bool = False) -> MultiTe
                 c.add_single(x, axy.mv(by, semiring).reshape(ishapes[x]))
     return c
 
+def _order_nonterminals(a: MultiTensor) -> list:
+    """ Reorders the nonterminals in a Multitensor.
+        Uses the heuristic found in Nederhof and Satta (2008) to put the 
+        linking nonterminals first, speeding up Gaussian Elimination
+    """    
+    if not a.keys():
+        return []
+    
+    #set up graph
+    nonterminal_graph: Dict[MultiTensorKey, set] = {}  
+    for x, y in a.keys():
+        if x not in nonterminal_graph:
+            nonterminal_graph[x] = set()
+            
+        nonterminal_graph[x].add(y)
+    
+    #dfs on graph to get finish times
+    finish_times = {}
+    visited = set()
+    time = [0] #this must be a list so it is mutable and will persist over multiple calls to dfs
```

Comment:
You can just make this an int, i.e., `time = 0`, then inside `dfs`, write `nonlocal time`, then to increment `time`, write `time += 1`